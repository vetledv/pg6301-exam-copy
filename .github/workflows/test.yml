name: Build

on:
  push:
    branches:
      - master

jobs:
  test:
    runs-on: ubuntu-latest
    timeout-minutes: 2
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: 16.x
      - name: install dependencies
        run: yarn install --frozen-lockfile
      - run: yarn test
        env:
          MONGO_URL: ${{ secrets.MONGO_URL }}
      - name: Coverage client
        uses: ArtiomTr/jest-coverage-report-action@v2
        with:
          package-manager: yarn
          working-directory: client
      - name: Coverage server
        uses: ArtiomTr/jest-coverage-report-action@v2
        with:
          package-manager: yarn
          working-directory: server
        env:
          MONGO_URL: ${{ secrets.MONGO_URL }}
          GOOGLE_CLIENT_ID: ${{ secrets.GOOGLE_CLIENT_ID }}
          GOOGLE_CLIENT_SECRET: ${{ secrets.GOOGLE_CLIENT_SECRET }}
          AZURE_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
          COOKIE_SECRET: ${{ secrets.AZURE_CLIENT_ID }}
