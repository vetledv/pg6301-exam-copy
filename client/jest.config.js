module.exports = {
    preset: 'ts-jest',
    testEnvironment: 'jsdom',
    clearMocks: true,
    collectCoverageFrom: [
        '**/*.{jsx,tsx,ts,js}',
        '!dist/**',
        '!coverage/**',
        '!**.config.js',
        '!src/mocks/**',
    ],
    testMatch: ['**/__tests__/**.test.{jsx,tsx,ts,js}'],
    moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
    setupFilesAfterEnv: ['./jest.setup.ts'],
}
