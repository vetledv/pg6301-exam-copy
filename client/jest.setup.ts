import { setLogger } from 'react-query'
import { server } from './src/mocks/server'

beforeAll(() => {
    server.listen({
        onUnhandledRequest({ method, url }) {
            if (!url.pathname.startsWith('/api')) {
                throw new Error(`Unhandled ${method} request to ${url}`)
            }
        },
    })
})
afterEach(async () => {
    server.resetHandlers()
})
afterAll(() => {
    server.close()
})

// suppress network errors being logged to the console for react-query
setLogger({
    log: console.log,
    warn: console.warn,
    error: () => {},
})

export const testArticle = {
    _id: '69',
    author: 'Vetle Brandth',
    body: 'Hello World',
    date: new Date('2022-05-06'),
    headline: 'Hello World',
    sub: 'dsrftygjuhrfty',
    topic: 'News',
}
export const { _id, author, body, topic } = testArticle

export const testUser = {
    config: {
        google: {
            response_type: 'token',
            authorization_endpoint:
                'https://accounts.google.com/o/oauth2/v2/auth',
            scope: 'email profile',
            userinfo_endpoint:
                'https://openidconnect.googleapis.com/v1/userinfo',
            client_id: process.env.GOOGLE_CLIENT_ID,
        },
        microsoft: {
            response_type: 'code',
            response_mode: 'fragment',
            client_id: process.env.AZURE_CLIENT_ID,
            code_challenge_method: 'S256',
            scope: 'openid User.Read',
            authorization_endpoint:
                'https://login.microsoftonline.com/organizations/oauth2/v2.0/authorize',
            token_endpoint:
                'https://login.microsoftonline.com/organizations/oauth2/v2.0/token',
            userinfo_endpoint: 'https://graph.microsoft.com/oidc/userinfo',
        },
    },
    user: {},
}
