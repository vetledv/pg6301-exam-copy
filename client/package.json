{
  "name": "client",
  "version": "1.0.0",
  "license": "MIT",
  "scripts": {
    "start": "yarn start",
    "dev": "parcel watch public/index.html",
    "build": "yarn install --production=false && yarn build:parcel",
    "build:parcel": "parcel build public/index.html",
    "test": "jest -c jest.config.js",
    "test:coverage": "jest -c jest.config.js --coverage",
    "verify": "tsc"
  },
  "dependencies": {
    "@parcel/transformer-react-refresh-wrap": "^2.5.0",
    "randomstring": "^1.2.2",
    "react": "^18.1.0",
    "react-dom": "^18.1.0",
    "react-icons": "^4.4.0",
    "react-query": "^3.39.1",
    "react-router-dom": "^6.3.0"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@testing-library/react-hooks": "^8.0.0",
    "@types/jest": "^27.4.1",
    "@types/pretty": "^2.0.1",
    "@types/randomstring": "^1.1.8",
    "@types/react": "^18.0.12",
    "@types/react-dom": "^18.0.5",
    "@types/react-router-dom": "^5.3.3",
    "jest": "^27.5.1",
    "jest-matchmedia-mock": "^1.1.0",
    "mongodb": "^4.5.0",
    "msw": "^0.42.1",
    "parcel": "^2.6.0",
    "postcss": "^8.4.13",
    "pretty": "^2.0.0",
    "process": "^0.11.10",
    "react-test-renderer": "^18.1.0",
    "tailwindcss": "^3.1.2",
    "ts-jest": "^27.1.4",
    "typescript": "^4.7.3",
    "whatwg-fetch": "^3.6.2"
  }
}