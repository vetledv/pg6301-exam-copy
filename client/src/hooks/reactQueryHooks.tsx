import { useMutation, useQuery, useQueryClient } from 'react-query'
import { useNavigate } from 'react-router-dom'
import { fetchJSON, postArticle, putJSON } from '../lib/fetch'

interface GoogleConfig {
    response_type: string
    authorization_endpoint: string
    scope: string
    userinfo_endpoint: string
    client_id: string
}

interface MicrosoftConfig extends GoogleConfig {
    response_mode: string
    code_challenge_method: string
    token_endpoint: string
}

interface MicrosoftUser {
    sub: string
    name: string
    family_name: string
    given_name: string
    picture: string
    email: string
}

interface GoogleUser extends MicrosoftUser {
    email_verified: boolean
    locale: string
}

export interface UseUser {
    config: {
        microsoft: MicrosoftConfig
        google: GoogleConfig
    }
    user: {
        google?: GoogleUser
        microsoft?: MicrosoftUser
    }
}

export interface UserNoConfig {
    google?: GoogleUser
    microsoft?: MicrosoftUser
}

export interface Article {
    _id?: string
    author: string
    body: string
    date: Date
    headline: string
    sub: string
    topic: string
}

export interface InfArticles {
    articles: Article[]
    next: number | null
}

export const useAllArticles = () => {
    return useQuery<Article[], Error>('articles', fetchJSON('/api/articles'))
}

export const useAllArticlesByTopic = (topic: string) => {
    return useQuery<Article[], Error>(
        ['articles', topic],
        fetchJSON('/api/articles/topic/' + topic)
    )
}

export const useArticle = (param: string) => {
    return useQuery<Article, Error>(
        ['article', param],
        fetchJSON('/api/articles/' + param)
    )
}

export const useMutateArticle = () => {
    const queryClient = useQueryClient()
    return useMutation((object: any) => postArticle(object), {
        onSuccess: () => {
            console.log('success')
            queryClient.refetchQueries('articles')
        },
    })
}

export const useEditArticle = (param: string) => {
    const queryClient = useQueryClient()
    return useMutation(
        (object: any) => putJSON(`/api/articles/${param}`, object),
        {
            onSuccess: () => {
                queryClient.refetchQueries('articles')
                queryClient.refetchQueries(['article', param])
            },
        }
    )
}

export const useUser = () => {
    return useQuery<UseUser, Error>('user', fetchJSON('/api/login'), {
        keepPreviousData: false,
        refetchOnReconnect: true,
    })
}

export const useMutateUser = (
    func: (provider?: string, object?: any) => Promise<void>,
    provider?: any
) => {
    const navigate = useNavigate()
    const queryClient = useQueryClient()
    return useMutation((object?: any) => func(provider, object), {
        onSuccess: () => {
            queryClient.refetchQueries('user')
            navigate('/')
        },
    })
}

//combined to one function useMutateUser, might need to be separate later

/*export const useLoginUser = (provider: any) => {
    const queryClient = useQueryClient()
    const navigate = useNavigate()
    return useMutation((object: any) => loginUser(provider, object), {
        onSettled: (data, variables) => {
            queryClient.refetchQueries('user')
            navigate('/')
        },
    })
}
export const useDeleteUser = () => {
    const navigate = useNavigate()
    const queryClient = useQueryClient()
    return useMutation((object) => logoutUser(), {
        onSettled: (data, variables) => {
            queryClient.refetchQueries('user')
            navigate('/')
        },
    })
}*/
