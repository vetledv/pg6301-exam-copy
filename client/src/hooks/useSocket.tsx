import { useEffect, useRef } from 'react'

const useSocket = (deps: any[] = [], callback: () => void) => {
    const socket = useRef<WebSocket | null>(null)

    useEffect(() => {
        if (!socket.current) return
        socket.current = new WebSocket(
            window.location.origin.replace(/^http/, 'ws')
        )

        socket.current.onopen = () => {
            console.log('socket opened')
        }
        socket.current.onclose = () => {
            console.log('socket closed')
        }
        socket.current.onerror = (e) => {
            console.log('socket error', e)
        }
        socket.current.onmessage = (event) => {
            console.log('deeznits')
            callback()
        }
    }, [...deps])

    return { socket }
}

export default useSocket
