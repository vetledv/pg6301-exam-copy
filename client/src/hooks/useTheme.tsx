import { Dispatch, SetStateAction, useEffect, useState } from 'react'

export type themeState = 'dark' | 'light'
export type setThemeState = Dispatch<SetStateAction<themeState>>
export interface ThemeProps {
    colorTheme: themeState
    setTheme: setThemeState
}

//default to user system pref
const mq = window.matchMedia('(prefers-color-scheme: dark)')
const userPref = mq.matches ? 'dark' : 'light'

export const useTheme = () => {
    const [theme, setTheme] = useState<themeState>(
        localStorage.theme || userPref
    )
    const colorTheme = theme === 'dark' ? 'light' : 'dark'

    useEffect(() => {
        const root = window.document.documentElement
        root.classList.remove(colorTheme)
        root.classList.add(theme)
        localStorage.setItem('theme', theme)
    }, [theme, colorTheme])
    return [colorTheme, setTheme] as const
}
