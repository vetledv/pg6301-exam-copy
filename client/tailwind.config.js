module.exports = {
    content: ['./src/**/*.{js,jsx,ts,tsx}', './public/index.html'],
    darkMode: 'class',
    theme: {
        screens: {
            sm: '640px',
            md: '768px',
            lg: '1024px',
            xl: '1280px',
            '2xl': '1536px',
        },
        extend: {
            backgroundColor: {
                primary: 'var(--color-bg-primary)',
                secondary: 'var(--color-bg-secondary)',
            },
            textColor: {
                accent: 'var(--color-text-accent)',
                primary: 'var(--color-text-primary)',
                secondary: 'var(--color-text-secondary)',
                contrast: 'var(--color-text-contrast)',
            },
            borderColor: {
                primary: 'var(--color-border-primary)',
                secondary: 'var(--color-border-secondary)',
            },
        },
        fontSize: {
            xs: '12px',
            sm: '14px',
            md: '16px',
            lg: '18px',
            xl: '24px',
            '2xl': '36px',
            '3xl': '48px',
        },
    },
    plugins: [],
}
