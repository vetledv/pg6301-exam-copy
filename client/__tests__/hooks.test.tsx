import { renderHook } from '@testing-library/react-hooks'
import { useAllArticles, useArticle } from '../src/hooks/reactQueryHooks'
import { createWrapper } from './client.test'

describe('custom hooks', () => {
    test('useAllArticles', async () => {
        const { result, waitFor } = renderHook(() => useAllArticles(), {
            wrapper: createWrapper(),
        })
        await waitFor(() => {
            return result.current.isSuccess
        })
        expect(result.current.isSuccess).toBe(true)
    })
    test('useArticle', async () => {
        const { result, waitFor } = renderHook(() => useArticle('69'), {
            wrapper: createWrapper(),
        })
        await waitFor(() => {
            return result.current.isSuccess
        })
        expect(result.current.isSuccess).toBe(true)
    })
})
