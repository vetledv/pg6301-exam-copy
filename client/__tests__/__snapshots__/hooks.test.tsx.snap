// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render Render without crashing 1`] = `"{\\"_reactInternalInstance\\":\\"censored\\"}"`;
